File: scripts/validate-package.py
"""Build and verify the exact Thunderstore ZIP used locally and in CI."""

import argparse
import json
import os
from pathlib import Path
import re
import struct
import sys
import xml.etree.ElementTree as ET
import zipfile


def metadata(root: Path):
    manifest = json.loads((root / "package/manifest.json").read_text())
    version = manifest["version_number"]
    if not re.fullmatch(r"(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)", version):
        raise ValueError("package/manifest.json must contain a three-part numeric version.")
    if any(int(part) > 65534 for part in version.split(".")):
        raise ValueError("Version components must be at most 65534 for .NET assembly versions.")
    if not re.fullmatch(r"[A-Za-z0-9_]+", manifest["name"]):
        raise ValueError("The package name must contain only letters, numbers, and underscores.")
    if not isinstance(manifest["description"], str) or len(manifest["description"]) > 250:
        raise ValueError("The package description must be at most 250 characters.")
    if not re.fullmatch(r"https://github\.com/[\w.-]+/[\w.-]+", manifest["website_url"]):
        raise ValueError("Set website_url to this mod's GitHub repository URL.")
    dependencies = manifest["dependencies"]
    if not isinstance(dependencies, list) or not all(
        isinstance(item, str) and re.fullmatch(r"\w+-\w+-\d+\.\d+\.\d+", item)
        for item in dependencies
    ):
        raise ValueError("Dependencies must use OWNER-PACKAGE-MAJOR.MINOR.PATCH.")
    if sum(item.startswith("denikson-BepInExPack_Valheim-") for item in dependencies) != 1:
        raise ValueError("Declare exactly one BepInExPack_Valheim dependency.")

    projects = list((root / "src").glob("*/*.csproj"))
    if len(projects) != 1:
        raise ValueError("Expected one plugin project at src/<Mod>/<Mod>.csproj.")
    project = projects[0]
    tree = ET.parse(project)
    if tree.findtext(".//Version") != version:
        raise ValueError("The project Version must match package/manifest.json. Run bumpp before tagging.")
    plugin = (project.parent / "Plugin.cs").read_text(encoding="utf-8-sig")
    versions = re.findall(r'public const string PluginVersion\s*=\s*"([^"]+)";', plugin)
    if versions != [version] or not re.search(r"\[BepInPlugin\([^\n]*, PluginVersion\)\]", plugin):
        raise ValueError("BepInPlugin must use PluginVersion, matching package/manifest.json.")
    assembly = (project.parent / "Properties/AssemblyInfo.cs").read_text(encoding="utf-8-sig")
    for attribute in ("AssemblyVersion", "AssemblyFileVersion"):
        values = re.findall(rf'^\[assembly: {attribute}\("([^"]+)"\)\]', assembly, re.MULTILINE)
        if values != [f"{version}.0"]:
            raise ValueError(f"{attribute} must match the package version with a .0 revision.")
    tag = os.environ.get("RELEASE_TAG", "")
    if tag and tag != f"v{version}":
        raise ValueError(f"Release tag {tag!r} must be 'v{version}'. No package was published.")

    icon = (root / "package/icon.png").read_bytes()
    if (icon[:8] != b"\x89PNG\r\n\x1a\n" or len(icon) < 24
            or struct.unpack(">II", icon[16:24]) != (256, 256)):
        raise ValueError("package/icon.png must be a 256 by 256 PNG.")
    name = tree.findtext(".//AssemblyName") or project.stem
    framework = tree.findtext(".//TargetFramework")
    dll = project.parent / f"bin/Release/{framework}/{name}.dll"
    archive = root / f"artifacts/{manifest['name']}-{version}.zip"
    return manifest, dll, archive


def package_files(root: Path, manifest, dll: Path):
    files = {str(path.relative_to(root / "package")): path.read_bytes()
             for path in sorted((root / "package").rglob("*")) if path.is_file()}
    if any(name.lower().endswith(".dll") for name in files):
        raise ValueError("Keep DLLs out of package/. Only the built plugin may be distributed.")
    readme = (root / "README.md").read_text()

    def link(match):
        prefix, target, suffix = match.groups()
        if target.startswith("package/"):
            target = target.removeprefix("package/")
        elif not re.match(r"(?:[a-z]+:|#|/)", target):
            target = f"{manifest['website_url']}/blob/main/{target}"
        return prefix + target + suffix

    # Keep one README source while making package assets and repository links resolve.
    readme = re.sub(r'(\]\()([^\s)]+)(\))', link, readme)
    readme = re.sub(r'(src=")([^"]+)(")', link, readme)
    files["README.md"] = readme.encode()
    for name in ("CHANGELOG.md", "LICENSE.md"):
        path = root / name
        if path.exists():
            files[name] = path.read_bytes()
    font_license = root / "assets/artwork/FONT-LICENSE.txt"
    if font_license.exists():
        files["FONT-LICENSE.txt"] = font_license.read_bytes()
    files[f"plugins/{dll.name}"] = dll.read_bytes()
    return files


def validate(root: Path, assembly: Path | None = None):
    manifest, dll, archive = metadata(root)
    if assembly is not None and assembly.resolve() != dll.resolve():
        raise ValueError("Package requires the Release build. Run dotnet build -c Release -t:Package.")
    expected = package_files(root, manifest, dll)
    if assembly is not None:
        archive.parent.mkdir(parents=True, exist_ok=True)
        temporary = archive.with_suffix(".zip.tmp")
        try:
            with zipfile.ZipFile(temporary, "w", zipfile.ZIP_DEFLATED) as package:
                for name, content in expected.items():
                    package.writestr(name, content)
            temporary.replace(archive)
        finally:
            temporary.unlink(missing_ok=True)
    with zipfile.ZipFile(archive) as package:
        if package.testzip() is not None:
            raise ValueError("ZIP integrity check failed. Rebuild the package.")
        if sorted(package.namelist()) != sorted(expected):
            raise ValueError("ZIP contents differ from the expected plugin and metadata. Rebuild the package.")
        for name, content in expected.items():
            if package.read(name) != content:
                raise ValueError(f"Packaged {name} differs from its source. Rebuild the package.")
    return manifest, archive


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--assembly", type=Path, help="Build a ZIP from this Release assembly before validating.")
    args = parser.parse_args()
    root = Path(__file__).resolve().parents[1]
    try:
        manifest, archive = validate(root, args.assembly)
    except (ValueError, KeyError, TypeError, OSError, ET.ParseError, zipfile.BadZipFile) as error:
        print(f"Package validation failed: {error}", file=sys.stderr)
        sys.exit(1)
    if output := os.environ.get("GITHUB_OUTPUT"):
        with open(output, "a") as stream:
            stream.write(f"version={manifest['version_number']}\n")
            stream.write(f"name={manifest['name']}\n")
            stream.write(f"archive={archive.relative_to(root)}\n")
    print(f"Validated {archive.relative_to(root)}.")

File: scripts/validate-package.ts
import assert from 'node:assert/strict';
import { appendFileSync, existsSync, mkdirSync, readFileSync, renameSync, rmSync, writeFileSync } from 'node:fs';
import { basename, dirname, join, relative, resolve } from 'node:path';
import { parseArgs } from 'node:util';
import { Glob, XML } from 'bun';
import { Uint8ArrayReader, Uint8ArrayWriter, ZipReader, ZipWriter } from '@zip.js/zip.js';

type PackageMode = { kind: 'check' } | { kind: 'build'; assembly: string };

function readManifest(root: string) {
  const value: unknown = JSON.parse(readFileSync(join(root, 'package/manifest.json'), 'utf8'));
  assert(typeof value === 'object' && value !== null, 'The package manifest must be an object.');
  assert('name' in value && typeof value.name === 'string' && /^[A-Za-z0-9_]+$/.test(value.name),
    'The package name must contain only letters, numbers, and underscores.');
  assert('version_number' in value && typeof value.version_number === 'string'
    && /^(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)$/.test(value.version_number)
    && value.version_number.split('.').every(part => Number(part) <= 65534),
  'The package version must use MAJOR.MINOR.PATCH with components between 0 and 65534.');
  assert('description' in value && typeof value.description === 'string' && value.description.length <= 250,
    'The package description must be at most 250 characters.');
  assert('website_url' in value && typeof value.website_url === 'string'
    && /^https:\/\/github\.com\/[\w.-]+\/[\w.-]+$/.test(value.website_url),
  "Set website_url to this mod's GitHub repository URL.");
  assert('dependencies' in value && Array.isArray(value.dependencies)
    && value.dependencies.every((item: unknown): item is string =>
      typeof item === 'string' && /^\w+-\w+-\d+\.\d+\.\d+$/.test(item)),
  'Dependencies must use OWNER-PACKAGE-MAJOR.MINOR.PATCH.');
  assert.equal(value.dependencies.filter(item => item.startsWith('denikson-BepInExPack_Valheim-')).length, 1,
    'Declare exactly one BepInExPack_Valheim dependency.');
  return { name: value.name, version: value.version_number, website: value.website_url };
}

function projectProperty(project: XML.Node, name: string): string | undefined {
  const values: string[] = [];
  for (const group of project.children) {
    if (typeof group === 'string' || !('name' in group) || group.name !== 'PropertyGroup') continue;
    for (const property of group.children) {
      if (typeof property === 'string' || !('name' in property) || property.name !== name) continue;
      assert(property.children.every(child => typeof child === 'string'), `Project ${name} must contain text.`);
      values.push(property.children.join('').trim());
    }
  }
  assert(values.length <= 1, `Expected at most one project ${name} property.`);
  return values[0];
}

function metadata(root: string, tag: string) {
  const manifest = readManifest(root);
  const projects = [...new Glob('src/*/*.csproj').scanSync({ cwd: root, absolute: true })];
  const [project] = projects;
  assert(project && projects.length === 1, 'Expected one plugin project at src/<Mod>/<Mod>.csproj.');
  const tree = XML.parse(readFileSync(project), { compact: false });
  assert.equal(projectProperty(tree, 'Version'), manifest.version,
    'The project Version must match package/manifest.json. Run bumpp before tagging.');
  const plugin = readFileSync(join(dirname(project), 'Plugin.cs'), 'utf8');
  assert.deepEqual([...plugin.matchAll(/public const string PluginVersion\s*=\s*"([^"]+)";/g)].map(match => match[1]),
    [manifest.version], 'PluginVersion must match package/manifest.json.');
  assert(/\[BepInPlugin\([^\n]*, PluginVersion\)\]/.test(plugin), 'BepInPlugin must use PluginVersion.');
  const assembly = readFileSync(join(dirname(project), 'Properties/AssemblyInfo.cs'), 'utf8');
  for (const attribute of ['AssemblyVersion', 'AssemblyFileVersion']) {
    const pattern = new RegExp(`^\\[assembly: ${attribute}\\("([^"]+)"\\)\\]`, 'gm');
    assert.deepEqual([...assembly.matchAll(pattern)].map(match => match[1]), [manifest.version + '.0'],
      `${attribute} must match the package version with a .0 revision.`);
  }
  assert(!tag || tag === `v${manifest.version}`,
    `Release tag ${JSON.stringify(tag)} must be v${manifest.version}. No package was published.`);
  const icon = readFileSync(join(root, 'package/icon.png'));
  assert(icon.length >= 24 && icon.subarray(0, 8).equals(Buffer.from('89504e470d0a1a0a', 'hex'))
    && icon.readUInt32BE(16) === 256 && icon.readUInt32BE(20) === 256,
  'package/icon.png must be a 256 by 256 PNG.');
  const name = projectProperty(tree, 'AssemblyName') ?? basename(project, '.csproj');
  const framework = projectProperty(tree, 'TargetFramework');
  assert(framework, 'The plugin project must declare TargetFramework.');
  return {
    ...manifest,
    dll: join(dirname(project), 'bin/Release', framework, name + '.dll'),
    archive: join(root, 'artifacts', `${manifest.name}-${manifest.version}.zip`),
  };
}

function packageFiles(root: string, info: ReturnType<typeof metadata>) {
  const files = new Map<string, Uint8Array>();
  const packageDir = join(root, 'package');
  for (const name of [...new Glob('**/*').scanSync({ cwd: packageDir, dot: true, onlyFiles: true })].sort()) {
    assert(!name.toLowerCase().endsWith('.dll'), 'Keep DLLs out of package/. Only the built plugin may be distributed.');
    files.set(name.replaceAll('\\', '/'), readFileSync(join(packageDir, name)));
  }
  function link(_match: string, prefix: string, target: string, suffix: string) {
    if (target.startsWith('package/')) {
      const repository = info.website.slice('https://github.com/'.length);
      target = `https://raw.githubusercontent.com/${repository}/v${info.version}/${target}`;
    }
    else if (!/^(?:[a-z]+:|#|\/)/.test(target)) target = `${info.website}/blob/main/${target}`;
    return prefix + target + suffix;
  }
  // Thunderstore does not serve bundled images; use the release tag's public files.
  const readme = readFileSync(join(root, 'README.md'), 'utf8')
    .replace(/(\]\()([^\s)]+)(\))/g, link).replace(/(src=")([^"]+)(")/g, link);
  files.set('README.md', Buffer.from(readme));
  for (const name of ['CHANGELOG.md', 'LICENSE.md', 'assets/artwork/FONT-LICENSE.txt']) {
    if (existsSync(join(root, name))) files.set(basename(name), readFileSync(join(root, name)));
  }
  files.set('plugins/' + basename(info.dll), readFileSync(info.dll));
  return files;
}

export async function validate(root: string, mode: PackageMode = { kind: 'check' }, tag = process.env.RELEASE_TAG ?? '') {
  const info = metadata(root, tag);
  if (mode.kind === 'build') {
    assert.equal(resolve(mode.assembly), resolve(info.dll),
      'Package requires the Release build. Run dotnet build -c Release -t:Package.');
  }
  const expected = packageFiles(root, info);
  if (mode.kind === 'build') {
    const writer = new ZipWriter(new Uint8ArrayWriter(), { useWebWorkers: false });
    for (const [name, bytes] of expected) await writer.add(name, new Uint8ArrayReader(bytes));
    const bytes = await writer.close();
    mkdirSync(dirname(info.archive), { recursive: true });
    const temporary = info.archive + '.tmp';
    try {
      writeFileSync(temporary, bytes);
      renameSync(temporary, info.archive);
    } finally {
      rmSync(temporary, { force: true });
    }
  }
  const reader = new ZipReader(new Uint8ArrayReader(readFileSync(info.archive)), { useWebWorkers: false });
  try {
    const entries = await reader.getEntries();
    assert.deepEqual(entries.map(entry => entry.filename).sort(), [...expected.keys()].sort(),
      'ZIP contents differ from the expected plugin and metadata. Rebuild the package.');
    for (const entry of entries) {
      assert(!entry.directory, 'Unexpected directory entry. Rebuild the package.');
      const bytes = await entry.getData(new Uint8ArrayWriter(), { checkSignature: true });
      const source = expected.get(entry.filename);
      assert(source && Buffer.from(bytes).equals(source),
        `Packaged ${entry.filename} differs from its source. Rebuild the package.`);
    }
  } finally {
    await reader.close();
  }
  return info;
}

if (import.meta.main) {
  try {
    const { values } = parseArgs({ options: { assembly: { type: 'string' } } });
    const root = resolve(import.meta.dir, '..');
    const info = await validate(root, values.assembly ? { kind: 'build', assembly: values.assembly } : { kind: 'check' });
    const archive = relative(root, info.archive).replaceAll('\\', '/');
    if (process.env.GITHUB_OUTPUT) {
      appendFileSync(process.env.GITHUB_OUTPUT, `version=${info.version}\nname=${info.name}\narchive=${archive}\n`);
    }
    console.log(`Validated ${archive}.`);
  } catch (error) {
    console.error(`Package validation failed: ${error instanceof Error ? error.message : String(error)}`);
    process.exitCode = 1;
  }
}
